File: Java/deepspace/src/deepspace/Weapon.java
/*
@authors Manuel Diaz-Meco Terrés y Ana Graciani Donaire
2ºDGIIM         PDOO


 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 * Esta clase representa a las armas de las que puede disponer una estación
 * espacial para potenciar su energía al disparar.
 */
public class Weapon implements CombatElement {
    
    private String name;
    private WeaponType type = null;
    private int uses;
    
    /**
     * Constructor con parámetros, visibilidad de paquete
     * @param name_
     * @param type_
     * @param uses_ 
     */
    public Weapon (String name_, WeaponType type_, int uses_) {
        name = name_;
        type = type_;
        uses = uses_;
    }
    
     /**
      * Constructor de copia
      * @param w Instancia Weapon a copiar
      */
    Weapon (Weapon w) {
        name = w.name;
        type = w.type;
        uses = w.uses;
    }

    /**
     * Consultor público del atributo type (tipo de arma)
     * @return type 
     */
    public WeaponType getType() {
        return type;
    }
    
    /**
     * Consultor público del atributo uses (usos del arma)
     * @return uses
     */
    @Override
    public int getUses() {
        return uses;
    }
    
    /**
     * Método que devuelve la potencia de disparo del arma
     * @return power indicado por el tipo de arma (type)
     */
    public float power() {
        return type.getPower();
    }
    
    /**
     * Si el valor del atributo uses es mayor que 0, lo decrementa en una
     * unidad y devuelve el valor del atributo boost
     * Devuelve el valor 1.0 en otro caso.
     */
    @Override
    public float useIt() {
        if (uses > 0) {
            uses--;
            return power();
        } else {
            return (float)1.0;
        }
    }
    
    /**
     * Método toString
     * @return String que representa el objeto Weapon
     */
    @Override
    public String toString() {
        
        return "name: " + name + ", type: " + type + ", uses: " + uses;
    }
    
    WeaponToUI getUIversion() {
        return (new WeaponToUI(this));
    }
}

File: Ruby/Weapon.rb
#encoding: UTF-8
#require_relative 'WeaponType.rb' # solo para el código de prueba

# Autores: Manuel Díaz-Meco y Ana Graciani Donaire
# 2º DGIIM		PDOO

# Esta clase representa a las armas de las que puede disponer una estación
# espacial para potenciar su energía al disparar.

require_relative 'WeaponToUI.rb'

module Deepspace

	class Weapon

		# Constructor con parámetros

		def initialize (name_, type_, uses_)
			# atributos privados de instancia
			@name = name_
			@type = type_
			@uses = uses_
		end

		# Método de copia (método de clase)

		def self.newCopy (w)
			new(w.name, w.type, w.uses)
		end

		public

		# Consultores

		# Consultor del atributo name
		def name
			return @name
		end

		# Consultor del atributo type
		def type
			return @type
		end

		# Consultor del atributo uses
		def uses
			return @uses
		end

		# Método que devuelve la potencia de disparo del arma según type
		def power
			return @type.power
		end

		# Método de uso del arma

		def useIt
			if (@uses > 0)
				@uses-= 1
				return power
			else
				return 1.0
			end
		end
		
		def to_s
			return "name: #{@name}, type: #{@type.name}, uses: #{@uses}"
		end
		
		def getUIversion
			return WeaponToUI.new(self)
		end
		
	end
end

# Código de prueba

=begin

	require_relative 'WeaponType.rb'

	arma1 = Deepspace::Weapon.new("manolo", Deepspace::WeaponType::LASER, 2)
	puts arma1.inspect
	puts arma1.name
	puts arma1.type
	puts arma1.uses
	puts arma1.power
	arma2 = Deepspace::Weapon.newCopy(arma1)
	puts arma2.inspect
	puts arma1.useIt
	puts arma1.useIt
	puts arma1.useIt
	arma3 = Deepspace::Weapon.new("kike", Deepspace::WeaponType::PLASMA, 0)
	puts arma3.useIt
	puts arma1.to_s
	puts arma1.getUIversion
=end
